File: src/cpp/cdm/io/protobuf/PBPatient.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "stdafx.h"
PUSH_PROTO_WARNINGS()
#include "pulse/cdm/bind/Patient.pb.h"
#include "pulse/cdm/bind/PatientNutrition.pb.h"
POP_PROTO_WARNINGS()
#include "io/protobuf/PBPatient.h"
#include "io/protobuf/PBProperties.h"
#include "io/protobuf/PBUtils.h"
#include "patient/SEPatient.h"
#include "properties/SEScalarTime.h"
#include "utils/FileUtils.h"

void PBPatient::Load(const CDM_BIND::PatientData& src, SEPatient& dst)
{
  dst.Clear();
  PBPatient::Serialize(src, dst);
}
void PBPatient::Serialize(const CDM_BIND::PatientData& src, SEPatient& dst)
{
  dst.SetName(src.name());
  dst.SetSex((ePatient_Sex)src.sex());
  if (src.has_age())
    PBProperty::Load(src.age(), dst.GetAge());
  if (src.has_weight())
    PBProperty::Load(src.weight(), dst.GetWeight());
  if (src.has_height())
    PBProperty::Load(src.height(), dst.GetHeight());
  if (src.has_bodydensity())
    PBProperty::Load(src.bodydensity(), dst.GetBodyDensity());
  if (src.has_bodyfatfraction())
    PBProperty::Load(src.bodyfatfraction(), dst.GetBodyFatFraction());
  if (src.has_leanbodymass())
    PBProperty::Load(src.leanbodymass(), dst.GetLeanBodyMass());
  if (src.has_idealbodyweight())
    PBProperty::Load(src.idealbodyweight(), dst.GetIdealBodyWeight());

  if (src.has_alveolisurfacearea())
    PBProperty::Load(src.alveolisurfacearea(), dst.GetAlveoliSurfaceArea());
  if (src.has_rightlungratio())
    PBProperty::Load(src.rightlungratio(), dst.GetRightLungRatio());
  if (src.has_skinsurfacearea())
    PBProperty::Load(src.skinsurfacearea(), dst.GetSkinSurfaceArea());

  if (src.has_basalmetabolicrate())
    PBProperty::Load(src.basalmetabolicrate(), dst.GetBasalMetabolicRate());
  if (src.has_bloodvolumebaseline())
    PBProperty::Load(src.bloodvolumebaseline(), dst.GetBloodVolumeBaseline());
  if (src.has_diastolicarterialpressurebaseline())
    PBProperty::Load(src.diastolicarterialpressurebaseline(), dst.GetDiastolicArterialPressureBaseline());
  if (src.has_heartratebaseline())
    PBProperty::Load(src.heartratebaseline(), dst.GetHeartRateBaseline());
  if (src.has_meanarterialpressurebaseline())
    PBProperty::Load(src.meanarterialpressurebaseline(), dst.GetMeanArterialPressureBaseline());
  if (src.has_respirationratebaseline())
    PBProperty::Load(src.respirationratebaseline(), dst.GetRespirationRateBaseline());
  if (src.has_systolicarterialpressurebaseline())
    PBProperty::Load(src.systolicarterialpressurebaseline(), dst.GetSystolicArterialPressureBaseline());
  if (src.has_tidalvolumebaseline())
    PBProperty::Load(src.tidalvolumebaseline(), dst.GetTidalVolumeBaseline());

  if (src.has_heartratemaximum())
    PBProperty::Load(src.heartratemaximum(), dst.GetHeartRateMaximum());
  if (src.has_heartrateminimum())
    PBProperty::Load(src.heartrateminimum(), dst.GetHeartRateMinimum());
  if (src.has_expiratoryreservevolume())
    PBProperty::Load(src.expiratoryreservevolume(), dst.GetExpiratoryReserveVolume());
  if (src.has_functionalresidualcapacity())
    PBProperty::Load(src.functionalresidualcapacity(), dst.GetFunctionalResidualCapacity());
  if (src.has_inspiratorycapacity())
    PBProperty::Load(src.inspiratorycapacity(), dst.GetInspiratoryCapacity());
  if (src.has_inspiratoryreservevolume())
    PBProperty::Load(src.inspiratoryreservevolume(), dst.GetInspiratoryReserveVolume());
  if (src.has_residualvolume())
    PBProperty::Load(src.residualvolume(), dst.GetResidualVolume());
  if (src.has_totallungcapacity())
    PBProperty::Load(src.totallungcapacity(), dst.GetTotalLungCapacity());
  if (src.has_vitalcapacity())
    PBProperty::Load(src.vitalcapacity(), dst.GetVitalCapacity());
}

CDM_BIND::PatientData* PBPatient::Unload(const SEPatient& src)
{
  CDM_BIND::PatientData* dst = new CDM_BIND::PatientData();
  PBPatient::Serialize(src, *dst);
  return dst;
}
void PBPatient::Serialize(const SEPatient& src, CDM_BIND::PatientData& dst)
{
  if (src.HasName())
    dst.set_name(src.m_Name);
  dst.set_sex((CDM_BIND::PatientData::eSex)src.m_Sex);
  if (src.HasAge())
    dst.set_allocated_age(PBProperty::Unload(*src.m_Age));
  if (src.HasWeight())
    dst.set_allocated_weight(PBProperty::Unload(*src.m_Weight));
  if (src.HasHeight())
    dst.set_allocated_height(PBProperty::Unload(*src.m_Height));
  if (src.HasBodyDensity())
    dst.set_allocated_bodydensity(PBProperty::Unload(*src.m_BodyDensity));
  if (src.HasBodyFatFraction())
    dst.set_allocated_bodyfatfraction(PBProperty::Unload(*src.m_BodyFatFraction));
  if (src.HasLeanBodyMass())
    dst.set_allocated_leanbodymass(PBProperty::Unload(*src.m_LeanBodyMass));
  if (src.HasIdealBodyWeight())
    dst.set_allocated_idealbodyweight(PBProperty::Unload(*src.m_IdealBodyWeight));

  if (src.HasAlveoliSurfaceArea())
    dst.set_allocated_alveolisurfacearea(PBProperty::Unload(*src.m_AlveoliSurfaceArea));
  if (src.HasRightLungRatio())
    dst.set_allocated_rightlungratio(PBProperty::Unload(*src.m_RightLungRatio));
  if (src.HasSkinSurfaceArea())
    dst.set_allocated_skinsurfacearea(PBProperty::Unload(*src.m_SkinSurfaceArea));

  if (src.HasBasalMetabolicRate())
    dst.set_allocated_basalmetabolicrate(PBProperty::Unload(*src.m_BasalMetabolicRate));
  if (src.HasBloodVolumeBaseline())
    dst.set_allocated_bloodvolumebaseline(PBProperty::Unload(*src.m_BloodVolumeBaseline));
  if (src.HasDiastolicArterialPressureBaseline())
    dst.set_allocated_diastolicarterialpressurebaseline(PBProperty::Unload(*src.m_DiastolicArterialPressureBaseline));
  if (src.HasHeartRateBaseline())
    dst.set_allocated_heartratebaseline(PBProperty::Unload(*src.m_HeartRateBaseline));
  if (src.HasMeanArterialPressureBaseline())
    dst.set_allocated_meanarterialpressurebaseline(PBProperty::Unload(*src.m_MeanArterialPressureBaseline));
  if (src.HasRespirationRateBaseline())
    dst.set_allocated_respirationratebaseline(PBProperty::Unload(*src.m_RespirationRateBaseline));
  if (src.HasSystolicArterialPressureBaseline())
    dst.set_allocated_systolicarterialpressurebaseline(PBProperty::Unload(*src.m_SystolicArterialPressureBaseline));
  if (src.HasTidalVolumeBaseline())
    dst.set_allocated_tidalvolumebaseline(PBProperty::Unload(*src.m_TidalVolumeBaseline));

  if (src.HasHeartRateMaximum())
    dst.set_allocated_heartratemaximum(PBProperty::Unload(*src.m_HeartRateMaximum));
  if (src.HasHeartRateMinimum())
    dst.set_allocated_heartrateminimum(PBProperty::Unload(*src.m_HeartRateMinimum));

  if (src.HasExpiratoryReserveVolume())
    dst.set_allocated_expiratoryreservevolume(PBProperty::Unload(*src.m_ExpiratoryReserveVolume));
  if (src.HasFunctionalResidualCapacity())
    dst.set_allocated_functionalresidualcapacity(PBProperty::Unload(*src.m_FunctionalResidualCapacity));
  if (src.HasInspiratoryCapacity())
    dst.set_allocated_inspiratorycapacity(PBProperty::Unload(*src.m_InspiratoryCapacity));
  if (src.HasInspiratoryReserveVolume())
    dst.set_allocated_inspiratoryreservevolume(PBProperty::Unload(*src.m_InspiratoryReserveVolume));
  if (src.HasResidualVolume())
    dst.set_allocated_residualvolume(PBProperty::Unload(*src.m_ResidualVolume));
  if (src.HasTotalLungCapacity())
    dst.set_allocated_totallungcapacity(PBProperty::Unload(*src.m_TotalLungCapacity));
  if (src.HasVitalCapacity())
    dst.set_allocated_vitalcapacity(PBProperty::Unload(*src.m_VitalCapacity));
}
void PBPatient::Copy(const SEPatient& src, SEPatient& dst)
{
  dst.Clear();
  CDM_BIND::PatientData data;
  PBPatient::Serialize(src, data);
  PBPatient::Serialize(data, dst);
}

bool PBPatient::SerializeToString(const SEPatient& src, std::string& output, SerializationFormat m)
{
  CDM_BIND::PatientData data;
  PBPatient::Serialize(src, data);
  return PBUtils::SerializeToString(data, output, m, src.GetLogger());
}
bool PBPatient::SerializeToFile(const SEPatient& src, const std::string& filename)
{
  CDM_BIND::PatientData data;
  PBPatient::Serialize(src, data);
  return PBUtils::SerializeToFile(data, filename, src.GetLogger());
}

bool PBPatient::SerializeFromString(const std::string& src, SEPatient& dst, SerializationFormat m)
{
  CDM_BIND::PatientData data;
  if (!PBUtils::SerializeFromString(src, data, m, dst.GetLogger()))
    return false;
  PBPatient::Load(data, dst);
  return true;
}
bool PBPatient::SerializeFromFile(const std::string& filename, SEPatient& dst)
{
  CDM_BIND::PatientData data;
  if (!PBUtils::SerializeFromFile(filename, data, dst.GetLogger()))
    return false;
  PBPatient::Load(data, dst);
  return true;
}

File: src/csharp/pulse/cdm/io/PBPatient.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

using System.Collections.Generic;
using Google.Protobuf;

namespace Pulse.CDM
{
  public class PBPatient
  {
    public static void Load(pulse.cdm.bind.PatientData src, SEPatient dst)
    {
      Serialize(src, dst);
    }
    public static void Serialize(pulse.cdm.bind.PatientData src, SEPatient dst)
    {
      dst.Clear();
      dst.SetName(src.Name);
      dst.SetSex((ePatient_Sex)(int)src.Sex);
      if (src.Age != null)
        PBProperty.Load(src.Age, dst.GetAge());
      if (src.Weight != null)
        PBProperty.Load(src.Weight, dst.GetWeight());
      if (src.Height != null)
        PBProperty.Load(src.Height, dst.GetHeight());
      if (src.BodyDensity != null)
        PBProperty.Load(src.BodyDensity, dst.GetBodyDensity());
      if (src.BodyFatFraction != null)
        PBProperty.Load(src.BodyFatFraction, dst.GetBodyFatFraction());
      if (src.IdealBodyWeight != null)
        PBProperty.Load(src.IdealBodyWeight, dst.GetIdealBodyWeight());
      if (src.LeanBodyMass != null)
        PBProperty.Load(src.LeanBodyMass, dst.GetLeanBodyMass());

      if (src.AlveoliSurfaceArea != null)
        PBProperty.Load(src.AlveoliSurfaceArea, dst.GetAlveoliSurfaceArea());
      if (src.RightLungRatio != null)
        PBProperty.Load(src.RightLungRatio, dst.GetRightLungRatio());
      if (src.SkinSurfaceArea != null)
        PBProperty.Load(src.SkinSurfaceArea, dst.GetSkinSurfaceArea());

      if (src.BasalMetabolicRate != null)
        PBProperty.Load(src.BasalMetabolicRate, dst.GetBasalMetabolicRate());
      if (src.BloodVolumeBaseline != null)
        PBProperty.Load(src.BloodVolumeBaseline, dst.GetBloodVolumeBaseline());
      if (src.DiastolicArterialPressureBaseline != null)
        PBProperty.Load(src.DiastolicArterialPressureBaseline, dst.GetDiastolicArterialPressureBaseline());
      if (src.HeartRateBaseline != null)
        PBProperty.Load(src.HeartRateBaseline, dst.GetHeartRateBaseline());
      if (src.MeanArterialPressureBaseline != null)
        PBProperty.Load(src.MeanArterialPressureBaseline, dst.GetMeanArterialPressureBaseline());
      if (src.RespirationRateBaseline != null)
        PBProperty.Load(src.RespirationRateBaseline, dst.GetRespirationRateBaseline());
      if (src.SystolicArterialPressureBaseline != null)
        PBProperty.Load(src.SystolicArterialPressureBaseline, dst.GetSystolicArterialPressureBaseline());
      if (src.TidalVolumeBaseline != null)
        PBProperty.Load(src.TidalVolumeBaseline, dst.GetTidalVolumeBaseline());

      if (src.HeartRateMaximum != null)
        PBProperty.Load(src.HeartRateMaximum, dst.GetHeartRateMaximum());
      if (src.HeartRateMinimum != null)
        PBProperty.Load(src.HeartRateMinimum, dst.GetHeartRateMinimum());
      if (src.ExpiratoryReserveVolume != null)
        PBProperty.Load(src.ExpiratoryReserveVolume, dst.GetExpiratoryReserveVolume());
      if (src.FunctionalResidualCapacity != null)
        PBProperty.Load(src.FunctionalResidualCapacity, dst.GetFunctionalResidualCapacity());
      if (src.InspiratoryCapacity != null)
        PBProperty.Load(src.InspiratoryCapacity, dst.GetInspiratoryCapacity());
      if (src.InspiratoryReserveVolume != null)
        PBProperty.Load(src.InspiratoryReserveVolume, dst.GetInspiratoryReserveVolume());
      if (src.ResidualVolume != null)
        PBProperty.Load(src.ResidualVolume, dst.GetResidualVolume());
      if (src.TotalLungCapacity != null)
        PBProperty.Load(src.TotalLungCapacity, dst.GetTotalLungCapacity());
      if (src.VitalCapacity != null)
        PBProperty.Load(src.VitalCapacity, dst.GetVitalCapacity());
    }
    public static bool SerializeFromString(string src, SEPatient dst)
    {
      try
      {
        pulse.cdm.bind.PatientData data = JsonParser.Default.Parse<pulse.cdm.bind.PatientData>(src);
        PBPatient.Load(data, dst);
      }
      catch (Google.Protobuf.InvalidJsonException)
      {
        return false;
      }
      return true;
    }

    public static pulse.cdm.bind.PatientData Unload(SEPatient src)
    {
      pulse.cdm.bind.PatientData dst = new pulse.cdm.bind.PatientData();
      Serialize(src, dst);
      return dst;
    }
    public static void Serialize(SEPatient src, pulse.cdm.bind.PatientData dst)
    {
      if (src.HasName())
        dst.Name = src.GetName();
      dst.Sex = (pulse.cdm.bind.PatientData.Types.eSex)(int)src.GetSex();
      if (src.HasAge())
        dst.Age = PBProperty.Unload(src.GetAge());
      if (src.HasWeight())
        dst.Weight = PBProperty.Unload(src.GetWeight());
      if (src.HasHeight())
        dst.Height = PBProperty.Unload(src.GetHeight());
      if (src.HasBodyDensity())
        dst.BodyDensity = PBProperty.Unload(src.GetBodyDensity());
      if (src.HasBodyFatFraction())
        dst.BodyFatFraction = PBProperty.Unload(src.GetBodyFatFraction());
      if (src.HasIdealBodyWeight())
        dst.IdealBodyWeight = PBProperty.Unload(src.GetIdealBodyWeight());
      if (src.HasLeanBodyMass())
        dst.LeanBodyMass = PBProperty.Unload(src.GetLeanBodyMass());

      if (src.HasAlveoliSurfaceArea())
        dst.AlveoliSurfaceArea = PBProperty.Unload(src.GetAlveoliSurfaceArea());
      if (src.HasRightLungRatio())
        dst.RightLungRatio = PBProperty.Unload(src.GetRightLungRatio());
      if (src.HasSkinSurfaceArea())
        dst.SkinSurfaceArea = PBProperty.Unload(src.GetSkinSurfaceArea());

      if (src.HasBasalMetabolicRate())
        dst.BasalMetabolicRate = PBProperty.Unload(src.GetBasalMetabolicRate());
      if (src.HasBloodVolumeBaseline())
        dst.BloodVolumeBaseline = PBProperty.Unload(src.GetBloodVolumeBaseline());
      if (src.HasDiastolicArterialPressureBaseline())
        dst.DiastolicArterialPressureBaseline = PBProperty.Unload(src.GetDiastolicArterialPressureBaseline());
      if (src.HasHeartRateBaseline())
        dst.HeartRateBaseline = PBProperty.Unload(src.GetHeartRateBaseline());
      if (src.HasMeanArterialPressureBaseline())
        dst.MeanArterialPressureBaseline = PBProperty.Unload(src.GetMeanArterialPressureBaseline());
      if (src.HasRespirationRateBaseline())
        dst.RespirationRateBaseline = PBProperty.Unload(src.GetRespirationRateBaseline());
      if (src.HasSystolicArterialPressureBaseline())
        dst.SystolicArterialPressureBaseline = PBProperty.Unload(src.GetSystolicArterialPressureBaseline());
      if (src.HasTidalVolumeBaseline())
        dst.TidalVolumeBaseline = PBProperty.Unload(src.GetTidalVolumeBaseline());

      if (src.HasHeartRateMaximum())
        dst.HeartRateMaximum = PBProperty.Unload(src.GetHeartRateMaximum());
      if (src.HasHeartRateMinimum())
        dst.HeartRateMinimum = PBProperty.Unload(src.GetHeartRateMinimum());
      if (src.HasExpiratoryReserveVolume())
        dst.ExpiratoryReserveVolume = PBProperty.Unload(src.GetExpiratoryReserveVolume());
      if (src.HasFunctionalResidualCapacity())
        dst.FunctionalResidualCapacity = PBProperty.Unload(src.GetFunctionalResidualCapacity());
      if (src.HasInspiratoryCapacity())
        dst.InspiratoryCapacity = PBProperty.Unload(src.GetInspiratoryCapacity());
      if (src.HasInspiratoryReserveVolume())
        dst.InspiratoryReserveVolume = PBProperty.Unload(src.GetInspiratoryReserveVolume());
      if (src.HasResidualVolume())
        dst.ResidualVolume = PBProperty.Unload(src.GetResidualVolume());
      if (src.HasTotalLungCapacity())
        dst.TotalLungCapacity = PBProperty.Unload(src.GetTotalLungCapacity());
      if (src.HasVitalCapacity())
        dst.VitalCapacity = PBProperty.Unload(src.GetVitalCapacity());

    }
    public static string SerializeToString(SEPatient src)
    {
      var pb = PBPatient.Unload(src);
      return pb.ToString();
    }
  }
}
